persist/sqlite/consensus_refactored_test.go

```diff
@@ -0,0 +1,943 @@
+package sqlite_test
+
+import (
+	"errors"
+	"math"
+	"path/filepath"
+	"testing"
+	"time"
+
+	"go.sia.tech/core/consensus"
+	"go.sia.tech/core/types"
+	"go.sia.tech/coreutils"
+	"go.sia.tech/coreutils/chain"
+	ctestutil "go.sia.tech/coreutils/testutil"
+	"go.sia.tech/explored/explorer"
+	"go.sia.tech/explored/internal/testutil"
+	"go.sia.tech/explored/persist/sqlite"
+	"go.uber.org/zap/zaptest"
+)
+
+func newStoreRefactored(t *testing.T, v2 bool, f func(*consensus.Network, types.Block)) (types.Block, consensus.State, *chain.DBStore, explorer.Store) {
+	log := zaptest.NewLogger(t)
+	dir := t.TempDir()
+
+	db, err := sqlite.OpenDatabase(filepath.Join(dir, "explored.sqlite3"), log.Named("sqlite3"))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	bdb, err := coreutils.OpenBoltChainDB(filepath.Join(dir, "consensus.db"))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	var network *consensus.Network
+	var genesisBlock types.Block
+	if v2 {
+		network, genesisBlock = ctestutil.V2Network()
+	} else {
+		network, genesisBlock = ctestutil.Network()
+	}
+	if f != nil {
+		f(network, genesisBlock)
+	}
+
+	store, genesisState, err := chain.NewDBStore(bdb, network, genesisBlock)
+	if err != nil {
+		t.Fatal(err)
+	}
+	bs := consensus.V1BlockSupplement{Transactions: make([]consensus.V1TransactionSupplement, len(genesisBlock.Transactions))}
+	_, au := applyUpdate(t, network.GenesisState(), bs, genesisBlock)
+	if err := db.UpdateChainState(nil, []chain.ApplyUpdate{au}); err != nil {
+		t.Fatal(err)
+	}
+
+	t.Cleanup(func() {
+		db.Close()
+		bdb.Close()
+	})
+	return genesisBlock, genesisState, store, db
+}
+
+func applyUpdate(t *testing.T, cs consensus.State, bs consensus.V1BlockSupplement, b types.Block) (consensus.State, chain.ApplyUpdate) {
+	if cs.Index.Height != math.MaxUint64 {
+		// don't validate genesis block
+		if err := consensus.ValidateBlock(cs, b, bs); err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	cs, au := consensus.ApplyBlock(cs, b, bs, time.Time{})
+	return cs, chain.ApplyUpdate{
+		ApplyUpdate: au,
+		Block:       b,
+		State:       cs,
+	}
+}
+
+func revertUpdate(t *testing.T, prevState consensus.State, bs consensus.V1BlockSupplement, b types.Block) chain.RevertUpdate {
+	ru := consensus.RevertBlock(prevState, b, bs)
+	return chain.RevertUpdate{
+		RevertUpdate: ru,
+		Block:        b,
+		State:        prevState,
+	}
+}
+
+func TestSiacoinOutput(t *testing.T) {
+	pk1 := types.GeneratePrivateKey()
+	addr1 := types.StandardUnlockHash(pk1.PublicKey())
+
+	pk2 := types.GeneratePrivateKey()
+	addr2 := types.StandardUnlockHash(pk2.PublicKey())
+
+	genesisBlock, cs, store, db := newStoreRefactored(t, false, func(network *consensus.Network, genesisBlock types.Block) {
+		genesisBlock.Transactions[0].SiacoinOutputs[0].Address = addr1
```

Comment:
The callback isn't great. I'd either double-down on the functional argument pattern and add options like `WithGenesisPayoutAddr` or pass the `network` and `genesisBlock` in as regular arguments. You could then have separate helpers like `v1TestNetwork(payoutAddr types.Address)` that you use to create the block and network whenever you need a specific payout address.